src/Npgsql/ConnectorPool.Multiplexing.cs

```diff
@@ -0,0 +1,459 @@
+                               writtenSynchronously &&
+                               _multiplexCommandReader.TryRead(out command))
+                        {
+                            Interlocked.Increment(ref connector.CommandsInFlightCount);
+                            writtenSynchronously = WriteCommand(connector, command, ref stats);
+                        }
+                    }
+                    else
+                    {
+                        // We reuse the timeout's cancellation token source as long as it hasn't fired, but once it has
+                        // there's no way to reset it (see https://github.com/dotnet/runtime/issues/4694)
+                        var timeoutTimeSpan = TimeSpan.FromTicks(timeout);
+                        timeoutTokenSource.CancelAfter(timeoutTimeSpan);
+                        if (timeoutTokenSource.IsCancellationRequested)
+                        {
+                            timeoutTokenSource.Dispose();
+                            timeoutTokenSource = new CancellationTokenSource(timeoutTimeSpan);
+                            timeoutToken = timeoutTokenSource.Token;
+                        }
+
+                        try
+                        {
+                            while (connector.WriteBuffer.WritePosition < _writeCoalescingBufferThresholdBytes &&
+                                   writtenSynchronously)
+                            {
+                                if (!_multiplexCommandReader.TryRead(out command))
+                                {
+                                    _waitsForFurtherCommands++;
+                                    command = await _multiplexCommandReader.ReadAsync(timeoutToken);
+                                }
+
+                                Interlocked.Increment(ref connector.CommandsInFlightCount);
+                                writtenSynchronously = WriteCommand(connector, command, ref stats);
+                            }
+
+                            // The cancellation token (presumably!) has not fired, reset its timer so
+                            // we can reuse the cancellation token source instead of reallocating
+                            timeoutTokenSource.CancelAfter(int.MaxValue);
+
+                            // Increase the timeout slightly for next time: we're under load, so allow more
+                            // commands to get coalesced into the same packet (up to the hard limit)
+                            timeout = Math.Min(timeout + WriteCoalescineDelayAdaptivityUs, _writeCoalescingDelayTicks);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // Timeout fired, we're done writing.
+                            // Reduce the timeout slightly for next time: we're under little load, so reduce impact
```

Comment:
Yeah, I had some fun in my head with this at some point, but let's see how this works in the world.

FYI in TE Fortunes I specify a timeout of 100us, though all batches complete within 60-70us (so the timeout never fires). Last time I checked, without the timeout perf was significantly less, since there are "hiccups" where no command is immediately available in the command channel; so the timeout really is to smooth over those.

Reply:
Do we care if we regress on normal load? TE is of course a pathological case.

Reply:
Multiplexing is definitely not going to be the right choice perf-wise in all scenarios - it's not something to be blindly turned on everywhere. So it's definitely an opt-in to be tried and tuned (with the various knobs), and if it helps - great. One good example is the head-of-line blocking problem - with multiplexing user code must process responses quickly, otherwise they're blocking the results of later queries on the same connection. We should consider introducing (optional) resultset buffering to mitigate this, but that would potentially have a memory cost for big resultsets... Just to say this isn't a simple problem unfortunately, and turning on multiplexing may regress in various scenarios.

Apart from that, of course we shouldn't regress in any significant way without multiplexing, for normal scenarios...